File: code/ex845.cpp
// Two pointers

class Solution {
public:
    int longestMountain(vector<int>& A) {
        int n = A.size();    if (n < 3)    return 0;
        
        int MAX = 0;
        int start = 0, end = 0;
        
        while (start < n) {
            // Find the start
            if (start + 1 >= n || A[start] >= A[start + 1]) {
                start++;
                continue;
            }
            end = start;
            
            // Find peak
            while (end + 1 < n && A[end] < A[end + 1])    end++;
            
            // Find the end point
            if (end + 1 < n && A[end] > A[end + 1]) {
                while (end + 1 < n && A[end] > A[end + 1])    end++;
                MAX = max(MAX, end - start + 1);
            }
          
            start = end;
        }
        
        return MAX;
    }
};

File: code/ex845.java
// Left and right arrays

class Solution {
    public int longestMountain(int[] A) {
        int n = A.length;    if (n == 0)    return 0;
        
        int[] left = new int[n];    left[0] = 1;
        int[] right = new int[n];    right[n - 1] = 1;
        
        for (int i = 1; i < n; i++) {
            left[i] = A[i - 1] < A[i] ? left[i - 1] + 1 : 1;
            int ri = n - 1 - i;
            right[ri] = A[ri] > A[ri + 1] ? right[ri + 1] + 1 : 1; 
        }
        
        // Get result
        int max = 0;
        for (int i = 0; i < n; i++) {
            if (left[i] == 1 || right[i] == 1)    continue;
            max = Math.max(max, left[i] + right[i] - 1);
        }
        
        return max;
    }
}

// Left:  [1, 1, 2, 3, 1, 1, 2]
// Right: [2, 1, 1, 3, 2, 1, 1]

// Left:  [1, 1, 1]
// Right: [1, 1, 1]